contributors/design-proposals/storage/pv-to-rbd-mapping.md

```diff
@@ -0,0 +1,83 @@
+# RBD Volume to PV Mapping 
+
+Authors: krmayankk@
+
+### Problem
+
+The RBD Dynamic Provisioner currently generates rbd volume names which are random. The
+current implementation generates a UUID and the rbd image name becomes 
+image := fmt.Sprintf("kubernetes-dynamic-pvc-%s", uuid.NewUUID()). This RBD image name 
+is stored in the PV. The PV also has a reference to the PVC to which it binds.  The problem 
+with this approach is that if there is a catostrophic etcd data loss and all PV's are lost,
+there is no way to recover the mapping from RBD to PV. Although the RBD volumes for the customer
+still exist, we have no way to tell which rbd volumes belong to which customer.
+
+## Goal
+
+We want to make RBD image name generation more deterministic, so that in catastrophic 
+situations, we can derive the PVC name from rbd image name/metadata and allow customer
+the following options:
+- Backup RBD volume data for specific customers and hand them their copy before deleting 
+  the RBD volume. Without knowing from rbd image , which customers they belong to we 
+  cannot do this.
+- Create PV with the given RBD name and pre-bind it to the desired PVC so that customer
+  can get its data back.
+
+## Motivation
+
+We recently had an etcd data lost which resulted in loss of this rbd to pv mapping and there
+was no way to restore customer data. This proposal aims to store pvc name as metadata in the
+RBD image name so that in catostrophic scenarios, the mapping can be restored by just looking
+at the RBD's.
+
+## Proposal
+
+### Implementation Detail
+
+#### Current Implementation
+
+```go
+func (r *rbdVolumeProvisioner) Provision() (*v1.PersistentVolume, error) {
+...
+
+ // create random image name
+ image := fmt.Sprintf("kubernetes-dynamic-pvc-%s", uuid.NewUUID())
+ r.rbdMounter.Image = image
+```
+
+#### New Implementation
+
+```go
+import b64 "encoding/base64"
+...
+
+
+func (r *rbdVolumeProvisioner) Provision() (*v1.PersistentVolume, error) {
+...
+
+
+ // create RBD Image name 
+ image := fmt.Sprintf("kubernetes-dynamic-pvc-%s", uuid.NewUUID())
+
+ // Create a base64 encoding of the PVC Name
+ rbdImageName := b64.StdEncoding.EncodeToString([]byte(r.options.PVC.Name))
+
+ // Append the base64 encoding to the string `kubernetes-dynamic-pvc-`
+ rbdImageName = fmt.Sprintf("kubernetes-dynamic-pvc-%s", rbdImageName)
```

Comment:
I am a bit concerned exposing PV name to rbd image name, for the following reason:
- the image name could run out of limit (cc @dillaman for (krbd) image name length and validation)
- it misses the randomness. Two PVCs with the same name (but different namespaces) then collide on PV name.
- if the PVC name has special meaning (e.g. "my-highly-confidential-pvc"), the name is accidentally leaked. Admittedly this may not be a real use case though.

What about using special key/value pairs for the image, e.g. through `rbd image-meta set`? It doesn't solve the 3nd problem but it fixes the first two problems without regression.

Reply:
@rootfs Some older versions of librbd / krbd start having issues w/ names longer than 95 characters.

Reply:
thanks @dillaman

Reply:
@rootfs thanks for the pointer on image-meta , i think that looks much better solution since it doesnt change the existing rbd image name, allows for arbitrary key value to be stored and kinda hides the pvc information in the meta rather than making it obvious in the name. I just tried it and it works.  I have updated the proposal, PTAL